yt/yt/client/logging/dynamic_table_log_writer.cpp

```diff
@@ -0,0 +1,373 @@
+                Config_->HighBacklogWeightWatermark);
+        }
+
+        if (Suspended_) {
+            return 0;
+        }
+
+        auto guard = Guard(SpinLock_);
+
+        // This is necessary to avoid hitting the limit of allowed row modifications within a single transaction.
+        // Also, there is no upside from huge batches.
+        if (CurrentRowCount_ >= Config_->MaxBatchRowCount || std::ssize(CurrentBuffer_) >= Config_->MaxBatchWeight) {
+            RotateBuffer();
+        }
+
+        TStringOutput outputStream(CurrentBuffer_);
+        TCountingOutput countingOutputStream(&outputStream);
+        Formatter_->WriteFormatted(&countingOutputStream, event);
+
+        auto writtenBytes = static_cast<i64>(countingOutputStream.Counter());
+
+        BacklogWeight_.fetch_add(writtenBytes, std::memory_order::relaxed);
+        ++CurrentRowCount_;
+
+        return writtenBytes;
+    }
+
+    // This method is called *extremely* often: once every 40ms.
+    // We intentionally do not trigger any flush actions, because they are very likely to
+    // take longer than 40ms to execute.
+    // Using a periodic executor for flushes is debatable. Alternatively, we could have
+    // implemented an infinite loop of flush iterations. The current solution seems better
+    // from a batching perspective, particularly in cases when the logging rate is moderate.
+    void Flush() override
+    { }
+
+    void Reload() override
+    { }
+
+    ~TDynamicTableLogWriter()
+    {
+        YT_UNUSED_FUTURE(FlushExecutor_->Stop());
+    }
+
+private:
+    const TClientHolderPtr ClientHolder_;
+    const std::unique_ptr<ILogFormatter> Formatter_;
+    const TDynamicTableLogWriterConfigPtr Config_;
+    const IInvokerPtr Invoker_;
+    const TPeriodicExecutorPtr FlushExecutor_;
+    const TLogger Logger;
+
+    DECLARE_THREAD_AFFINITY_SLOT(LoggingThread);
+
+    //! Protects the fields below.
+    YT_DECLARE_SPIN_LOCK(NThreading::TSpinLock, SpinLock_);
+    TString CurrentBuffer_;
+    i64 CurrentRowCount_ = 0;
+    //! This queue should typically contain no more than one element.
+    //! We use it to limit the number of rows written within a single transaction.
+    std::deque<TString> BufferQueue_;
+
+    //! Only accessed in the logging thread.
+    bool Suspended_ = false;
+
+    //! Stored in terms of YSON bytes.
+    std::atomic<i64> BacklogWeight_{0};
+
+    void RotateBuffer()
+    {
+        VERIFY_SPINLOCK_AFFINITY(SpinLock_);
+
+        if (CurrentBuffer_.empty()) {
+            return;
+        }
+
+        BufferQueue_.push_back(std::move(CurrentBuffer_));
+        // Moved-from state should be fine, but just to be sure.
+        CurrentBuffer_.clear();
+        CurrentRowCount_ = 0;
+    }
+
+    // Periodic executor guarantees that at most one callback is executed at once.
+    // This is important to avoid log reordering.
+    void DoFlush()
+    {
+        VERIFY_INVOKER_AFFINITY(Invoker_);
+
+        TString ysonRows;
```

Comment:
Why TString, not TYsonString?
Or maybe even shared refs to ease memory tracking?

Reply:
Changed to `TSharedRef` for better memory tracking. But more generally speaking I would prefer to move towards the ability to pass yson input to C++ client write methods.

Reply:
Switched to shared refs.